pkg/resource/aggregator.go

```diff
@@ -0,0 +1,215 @@
+			var mem resource.Quantity
+
+			// read the container memory limits
+			for _, container := range nodeSet.PodTemplate.Spec.Containers {
+				if container.Name == esv1.ElasticsearchContainerName {
+					mem = *container.Resources.Limits.Memory()
+
+					// if not, fallback to twice the max JVM heap size
+					if mem.IsZero() {
+						for _, envVar := range container.Env {
+							if envVar.Name == essettings.EnvEsJavaOpts {
+								mem, err = memFromJavaOpts(envVar.Value)
+								if err != nil {
+									return resource.Quantity{}, err
+								}
+							}
+						}
+					}
+				}
+			}
+
+			// if not, fallback to the default limits
+			if mem.IsZero() {
+				mem = nodespec.DefaultMemoryLimits
+			}
+
+			total.Add(multiply(mem, nodeSet.Count))
+			log.V(1).Info("Collecting", "es_name", es.Name, "memory", mem.String(), "count", nodeSet.Count)
+		}
+	}
+
+	return total, nil
+}
+
+func (a Aggregator) aggregateKibanaMemory() (resource.Quantity, error) {
+	var kbList kbv1.KibanaList
+	err := a.client.List(&kbList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, kb := range kbList.Items {
+		var mem resource.Quantity
+
+		// read the container memory limits
+		for _, container := range kb.Spec.PodTemplate.Spec.Containers {
+			if container.Name == kbv1.KibanaContainerName {
+				mem = *container.Resources.Limits.Memory()
+
+				// if not, fallback to the max JVM heap size
+				if mem.IsZero() {
+					for _, envVar := range container.Env {
+						if envVar.Name == kbconfig.EnvNodeOpts {
+							mem, err = memFromNodeOptions(envVar.Value)
+							if err != nil {
+								return resource.Quantity{}, err
+							}
+						}
+					}
+				}
+			}
+		}
+
+		// if not, fallback to the default limits
+		if mem.IsZero() {
+			mem = kbpod.DefaultMemoryLimits
+		}
+
+		total.Add(multiply(mem, kb.Spec.Count))
+		log.V(1).Info("Collecting", "kibana_name", kb.Name, "memory", mem.String(), "count", kb.Spec.Count)
+	}
+
+	return total, nil
+}
+
+func (a Aggregator) aggregateApmServerMemory() (resource.Quantity, error) {
+	var asList asv1.ApmServerList
+	err := a.client.List(&asList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, as := range asList.Items {
+		var mem resource.Quantity
+
+		// read the container memory limits
+		for _, container := range as.Spec.PodTemplate.Spec.Containers {
+			if container.Name == asv1.ApmServerContainerName {
+				mem = *container.Resources.Limits.Memory()
+			}
+		}
+
+		// if not, fallback to the default limits
+		if mem.IsZero() {
+			mem = apmserver.DefaultMemoryLimits
+		}
+
+		total.Add(multiply(mem, as.Spec.Count))
+		log.V(1).Info("Collecting", "as_name", as.Name, "memory", mem.String(), "count", as.Spec.Count)
+	}
+
+	return total, nil
+}
+
+// maxHeapSizePattern is the pattern to extract the max Java heap size (-Xmx<size>[g|G|m|M|k|K])
+const maxHeapSizePattern = "-Xmx([0-9]*)([gGmMkK]*)"
```

Comment:
```suggestion
const maxHeapSizePattern = "-Xmx([0-9]*)([gGmMkK]?)$"
```

Otherwise we would match and try parse things like `5MM`.

Reply:
Yes for the `?` but no for the `$`. 
We don't know if -Xmx is the latest option declared in the JAVA_OPTS env var.

Reply:
Ah, got it. But then we should include space at the end, right? Also, should we do `[0-9]+` instead of `[0-9]*`? For cases: `-Xmx5GMoresettings` and `-XmxM`.

Reply:
> But then we should include space at the end, right?

No?! But we can be more precise and wait for only 1 character.

> Also, should we do `[0-9]+` instead of `[0-9]*`?

Yes!

Which gives: `"-Xmx([0-9]+)([gGmMkK])"`.

Reply:
> No?! But we can be more precise and wait for exactly 1 character.

What I meant is that in case of some badly formatted options (I'm not even sure that could be the case) we might match the pattern successfully and then fail on parsing. And the example could be `-Xmx5GMoresettings`, where we would match `GM` in second group. 

As to the `-Xmx([0-9]+)([gGmMkK])` I think the default unit is bytes, which won't have any letter, so we probably have to support that :)

I think the right one would be `-Xmx([0-9]+)([gGmMkK]?(?:\ .+|$))`. At the end we either want to see space followed by any other characters or just the end. Or we can accept failing on parsing in rare cases and just do simpler `-Xmx([0-9]+)([gGmMkK]?)`.